### PR TITLE
Add range get from list and list multiplication

### PR DESCRIPTION
**Pull request by MarcinMoskala**
There are to 2 handy features to get elements from list like:
list[0, 1, 2, 5] or list[2..4]

### CODE REVIEW

**Comment by cy6erGn0m** [comment]
Doesn't vararg getter conflict with the regular `get`?

**Comment by cy6erGn0m** [comment]
For me `get` with vararg is unnecessary thing but I like the `IntRange` idea

**Comment by cy6erGn0m** [comment]
But I am not sure about implementation: shouldn't we do `subList(range).toList()` instead?

**Comment by MarcinMoskala** [comment]
There is no conflict in java 
http://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.12.2.5

**Comment by MarcinMoskala** [comment]
More like:

operator fun <T> List<T>.get(indexesRange: IntRange): List<T> = subList(indexesRange.first, indexesRange.last)

?

**Comment by MarcinMoskala** [comment]
Let's do it then

**Comment by MarcinMoskala** [comment]
I correct it. Could You take a look on that? :)

**Comment by MarcinMoskala** [comment]
I also added list multiplication. Also usefull feature

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 1]
```diff
@@ -235,6 +235,22 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
 }
 
 /**
+ * Returns the elements from the list from indexes specified in range.
+ */
+operator fun <T> List<T>.get(indexesRange: IntRange): List<T> =
+        if(indexesRange.first < 0 || indexesRange.last > lastIndex)
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: What is the value of `last` of empty range and of `lastIndex` of the empty list? Don't they throw an exception? You have to add corresponding tests

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 2]
```diff
@@ -235,6 +235,22 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
...
+            throw IndexOutOfBoundsException()
+        else if(indexesRange.first<=indexesRange.last)
+            subList(indexesRange.first, indexesRange.last+1)
+        else
+            listOf<T>()
+
+/**
+ * Returns the list multiplication.
+ */
+operator fun <T> List<T>.times(factor: Int): List<T> = (1..factor).flatMap { this }
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Negative `factor` should throw an exception

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 3]
```diff
@@ -235,6 +235,22 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
 }
 
 /**
+ * Returns the elements from the list from indexes specified in range.
+ */
+operator fun <T> List<T>.get(indexesRange: IntRange): List<T> =
+        if(indexesRange.first < 0 || indexesRange.last > lastIndex)
+            throw IndexOutOfBoundsException()
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: It should be more debug information: exact range bounds and the list size

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 4]
```diff
@@ -235,6 +235,22 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
...
+        else if(indexesRange.first<=indexesRange.last)
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Formatting

File: libraries/stdlib/test/collections/CollectionTest.kt [hunk 5]
```diff
@@ -854,4 +854,38 @@ class CollectionTest {
...
+            f()
+        } catch(e: Exception){
+            return true
+        }
+        return false
+    }
+
+    @test fun testListTimes(){
+        val shortlistStr = listOf("a", "b")
+        assertTrue(shortlistStr*0 == listOf<String>())
```
Comment (cy6erGn0m) [inline libraries/stdlib/test/collections/CollectionTest.kt]: `assertEquals(expected, actual) ` is preferred as it dumps diff on failure

File: libraries/stdlib/src/kotlin/collections/Collections.kt (hunk 1 above)
Comment (MarcinMoskala) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Done

File: libraries/stdlib/src/kotlin/collections/Collections.kt (hunk 3 above)
Comment (MarcinMoskala) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Done

File: libraries/stdlib/src/kotlin/collections/Collections.kt (hunk 4 above)
Comment (MarcinMoskala) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Done

File: libraries/stdlib/test/collections/CollectionTest.kt (hunk 5 above)
Comment (MarcinMoskala) [inline libraries/stdlib/test/collections/CollectionTest.kt]: Done

File: libraries/stdlib/src/kotlin/collections/Collections.kt (hunk 2 above)
Comment (MarcinMoskala) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: This is controversive. I was thinking about it and I check how it works in Python ant there for minus factors there is no error but empyt list. But I will change it If you prefer this way.

Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Yes, our other functions do that so `times` should as well

File: libraries/stdlib/test/collections/CollectionTest.kt [hunk 6]
```diff
@@ -854,4 +854,41 @@ class CollectionTest {
...
         assertEquals("[1, a, null, ${Long.MAX_VALUE.toString()}]", listOf(1, "a", null, Long.MAX_VALUE).toString())
     }
+
+    @test fun testGetRange() {
+        val listStr = listOf("a", "b", "c", "d", "e")
+
+        assertTrue(throwsError{listStr[-1..2]})
+        assertTrue(throwsError{listStr[1..10]})
+
+        assertEquals(listStr[IntRange.EMPTY], listOf<String>())
```
Comment (cy6erGn0m) [inline libraries/stdlib/test/collections/CollectionTest.kt]: I should be expected value at the first place, the second argument is actual, swap these arguments please

File: libraries/stdlib/test/collections/CollectionTest.kt [hunk 7]
```diff
@@ -854,4 +854,41 @@ class CollectionTest {
         // we need toString() inside pattern because of KT-8666
         assertEquals("[1, a, null, ${Long.MAX_VALUE.toString()}]", listOf(1, "a", null, Long.MAX_VALUE).toString())
     }
+
+    @test fun testGetRange() {
+        val listStr = listOf("a", "b", "c", "d", "e")
```
Comment (cy6erGn0m) [inline libraries/stdlib/test/collections/CollectionTest.kt]: I believe we also need tests on an empty list: should fail at any range but empty range on empty list should produce empty list

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 8]
```diff
@@ -235,6 +235,28 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
...
+            subList(indexesRange.first, indexesRange.last + 1)
+        else
+            listOf<T>()
+
+/**
+ * Returns the list multiplication.
+ */
+operator fun <T> List<T>.times(factor: Int): List<T> =
+        if (factor < 0)
+            throw ArithmeticException("List factor must be bigger then 0")
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: You may just do `require(factor >= 0) {"error message… "}`. Also notice that in Java wrong arguments should cause `IllegalArgumentException ` (so it is done inside of the `require` function)

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 9]
```diff
@@ -235,6 +235,28 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
...
 /**
+ * Returns the elements from the list from indexes specified in range.
+ */
+operator fun <T> List<T>.get(indexesRange: IntRange): List<T> =
+        if (indexesRange.first < 0)
+            throw IndexOutOfBoundsException("Range beginning (${indexesRange.first}) is less than zero.")
+        else if (indexesRange.last > lastIndex)
+            throw IndexOutOfBoundsException("Range end (${indexesRange.last}) is bigger then the end of table ($lastIndex)")
+        else if (indexesRange.first <= indexesRange.last)
+            subList(indexesRange.first, indexesRange.last + 1)
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: It should be `toList()` after because we need a copy but `subList` just refers to the original list.

**Comment by cy6erGn0m** [comment]
Generally it is much better now, thank you for your help. Fix the remaining notes and we'll discuss it. I think it is the good one candidate for merge

File: libraries/stdlib/test/collections/CollectionTest.kt (hunk 6 above)
Comment (MarcinMoskala) [inline libraries/stdlib/test/collections/CollectionTest.kt]: Done

File: libraries/stdlib/test/collections/CollectionTest.kt (hunk 7 above)
Comment (MarcinMoskala) [inline libraries/stdlib/test/collections/CollectionTest.kt]: Done

File: libraries/stdlib/src/kotlin/collections/Collections.kt (hunk 8 above)
Comment (MarcinMoskala) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Done

File: libraries/stdlib/src/kotlin/collections/Collections.kt (hunk 9 above)
Comment (MarcinMoskala) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Done

**Comment by MarcinMoskala** [comment]
Fixed

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 10]
```diff
@@ -235,6 +235,30 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
...
 /**
+ * Returns the elements from the list from indexes specified in range.
+ */
+public operator fun <T> List<T>.get(indexesRange: IntRange): List<T> =
+        if (indexesRange.isEmpty())
+            listOf<T>()
+        else if (indexesRange.first < 0)
+            throw IndexOutOfBoundsException("Range beginning (${indexesRange.first}) is less than zero.")
+        else if (indexesRange.last > lastIndex)
+            throw IndexOutOfBoundsException("Range end (${indexesRange.last}) is bigger then the end of table ($lastIndex)")
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: typo: than

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 11]
```diff
@@ -235,6 +235,30 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
...
+        else if (indexesRange.first <= indexesRange.last)
+            subList(indexesRange.first, indexesRange.last + 1).toList()
+        else
+            throw Exception("Unknown exception in getting range of indexes from list")
+
+/**
+ * Returns the list multiplication.
+ */
+public operator fun <T> List<T>.times(factor: Int): List<T> {
+    require(factor >= 0) { "List factor must be bigger then 0" }
```
Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Type: than
Also lets include actual value in the message

Comment (cy6erGn0m) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: Also, in fact the message is not exact correct: factor shouldn't be negative so zero is also included

**Comment by HappyEmu** [comment]
I'd like to throw in the idea to support negative upper range limits as well. The negative upper limit would represent the offset from the size of the list. Example:

`list[2..-1]` would return all elements starting from index 2 until the end (`list.size - 1`) of the list.
`list[3..-2]` would return all elements starting from index 3 until index `list.size - 2`

**Comment by MarcinMoskala** [comment]
Can do it. And wanted to do it. Just don't know if rest agree with that. That is why I prefer to close it like that right now (because it is my first pull request and I care a lot) and add this funcionality in next pull request.

**Comment by cy6erGn0m** [comment]
@MarcinMoskala Another one problem I can see is `list[1..10 step 2]`. Open question is: `list[10 downTo 1] = ?`

**Comment by cy6erGn0m** [comment]
Regarding negative range bounds: not sure it is good idea as it may cause strange consequences. See

```
fun myfun(someRange: IntRange) = myList[someRange]
```

In this case we can potentially get unpredictable results as IntRange could be computed somehow

```
myfun(1..someString.indexOf(something))
```

**Comment by MarcinMoskala** [comment]
That is why 

operator fun <T> List<T>.get(indexesRange: IntRange): List<T> = indexesRange.toList().map { get(it) }

was perfect xP

**Comment by MarcinMoskala** [comment]
With - values:

List.get(indexesRange: IntRange): List = indexesRange.toList().map {if(i<0) i-size else i}. map { get(it) }

**Comment by cy6erGn0m** [comment]
Why do you need `toList()` here?
I still believe it is not good idea to support negative values

File: libraries/stdlib/src/kotlin/collections/Collections.kt [hunk 12]
```diff
@@ -237,24 +237,26 @@ public fun <T> List<T>.binarySearch(fromIndex: Int = 0, toIndex: Int = size, com
...
 /**
  * Returns the list multiplication.
  */
-operator fun <T> List<T>.times(factor: Int): List<T> =
-        if (factor < 0)
-            throw ArithmeticException("List factor must be bigger then 0")
-        else
-            (1..factor).flatMap { this }
+public operator fun <T> List<T>.times(factor: Int): List<T> {
+    require(factor >= 0) { "List factor must be bigger then 0" }
```
Comment (HaasJona) [inline libraries/stdlib/src/kotlin/collections/Collections.kt]: "than", also is 0 itself valid or not?
